### PR TITLE
Implemented Resource categories, subcategories, and item list screens

### PR DESCRIPTION
**Pull request by Derek-Chu**

### CODE REVIEW

File: src/screens/Resources/HomeScreen.js [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+            .where('parent', '==', id)
+            .get()
+            .then((itemSnapshot) => {
+              navigation.push(nav.itemList, { snapshot: itemSnapshot, header: title });
+            })
+            .catch((error) => {
+              setErrorMessage(error.message);
+              setLoading(false);
+            });
+        }
```
Comment (vfcheung) [inline src/screens/Resources/HomeScreen.js]: So this makes sense to me because we discussed how to design this feature, but someone new to this code might find it hard to parse these lines. Can we refactor? In fact `async` and `await` might help to make it easier to parse.

File: src/screens/Resources/HomeScreen.js [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+            return (
+              <TouchableOpacity
+                style={HomeStyles.categoryButton}
+                key={doc.id}
+                disabled={loading}
+                onPress={() => loadScreen(doc.id, title)}
+              >
+                <Image source={{ uri: thumbnail.src }} style={HomeStyles.categoryImage} />
+                <Text style={HomeStyles.categoryText}>
+                  {`${title}`}
```
Comment (vfcheung) [inline src/screens/Resources/HomeScreen.js]: You can just put `{title}`

File: src/screens/Resources/variables.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+export const collections = {
```
Comment (vfcheung) [inline src/screens/Resources/variables.js]: We can probably rename this file to `constants` and move it up so other components can use it.

File: src/screens/Resources/HomeScreen.js [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+                </Text>
+              </TouchableOpacity>
+            );
+          }))}
+      </View>
```
Comment (vfcheung) [inline src/screens/Resources/HomeScreen.js]: Might be worthwhile to create a separate component for this list so this component is less monolithic.

File: src/screens/Resources/ResourcesItemListScreen.js [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        <View style={ItemListStyles.moreButton}>
+          <Text
+            style={ItemListStyles.resourceText}
+            onPress={() => {
+              setLoading(true);
+              navigation.push(nav.item, { resourceID: doc.id });
+            }}
+          >
+            More
+            {' '}
```
Comment (vfcheung) [inline src/screens/Resources/ResourcesItemListScreen.js]: Use `&nbsp` (non-breaking space)

File: src/screens/Resources/ResourcesItemListScreen.js [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    );
+  }
+
+  function updateSections(currActive) {
+    setActiveSections(currActive);
+  }
+
+  function singleOrMultipleDocs() {
+    // returns a collapsible if there is only 1 doc
+    if (snapshot.docs.length === 1) {
```
Comment (vfcheung) [inline src/screens/Resources/ResourcesItemListScreen.js:20]: What if snapshot is empty?

File: src/screens/Resources/ResourcesItemListScreen.js [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+  const [loading, setLoading] = useState(false);
+  const [activeSections, setActiveSections] = useState([]);
+  const [singleActive, setSingleActive] = useState(false);
+
+  function renderHeader(doc, _, isActive) {
+    return (
+      <View style={ItemListStyles.accordionHeader}>
+        <View style={ItemListStyles.accordionHeaderText}>
+          <Text style={ItemListStyles.resourceText}>
+            {doc.data().title}
```
Comment (vfcheung) [inline src/screens/Resources/ResourcesItemListScreen.js]: In case we have to do null checks, calling `doc.data()` here might make that hard to do. Can we call `doc.data()` in one place, store the result pass it to these components?

File: src/screens/Resources/ResourcesItemListScreen.js [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+  function updateSections(currActive) {
+    setActiveSections(currActive);
+  }
+
+  function singleOrMultipleDocs() {
+    // returns a collapsible if there is only 1 doc
+    if (snapshot.docs.length === 1) {
+      const doc = snapshot.docs[0];
+      return (
+        <View>
```
Comment (vfcheung) [inline src/screens/Resources/ResourcesItemListScreen.js]: These components can be refactored into smaller components.

File: src/screens/Resources/ResourcesItemListScreen.js [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+        <View style={ItemListStyles.accordionHeaderIcon}>
+          <Icon name={isActive ? 'chevron-up' : 'chevron-down'} />
+        </View>
+      </View>
+    );
+  }
+
+  function renderContent(doc) {
+    return (
+      <View style={ItemListStyles.accordionContent}>
```
Comment (vfcheung) [inline src/screens/Resources/ResourcesItemListScreen.js]: This can be refactored into its own component

File: src/screens/Resources/ResourcesSubcategoriesScreen.js [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+import { collections, nav } from './variables';
+
+
+export default function ResourcesSubcategoriesScreen(props) {
+  const { navigation } = props;
+  const snapshot = navigation.getParam('snapshot');
+  const [loading, setLoading] = useState(false);
+  const [errorMessage, setErrorMessage] = useState(null);
+
+  const loadScreen = useCallback((id, title) => {
```
Comment (vfcheung) [inline src/screens/Resources/ResourcesSubcategoriesScreen.js]: Can we refactor this with the other `loadScreen` into one file?

File: src/screens/Resources/ResourcesSubcategoriesScreen.js [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+        && (
+        <Text style={{ color: 'red' }}>
+          {errorMessage}
+        </Text>
+        )}
+      {loading
+        && (
+        <ActivityIndicator size="large" />
+        )}
+      { snapshot.docs.map((doc) => {
```
Comment (vfcheung) [inline src/screens/Resources/ResourcesSubcategoriesScreen.js]: Can make this its own component

**Review by vfcheung — CHANGES_REQUESTED** [review]
Good work, let's work on these changes for now.

File: src/screens/Resources/ResourcesItemListScreen.js (hunk 6 above)
Comment (Derek-Chu) [reply src/screens/Resources/ResourcesItemListScreen.js:20]: if snapshot is empty it works fine, there's just no interactive parts to the screen.

File: android/app/src/main/res/raw/app.json [hunk 12]
```diff
@@ -0,0 +1,4 @@
+{
+  "name": "PTVMobileApp",
+  "displayName": "PTVMobileApp"
+}
```
Comment (enmingliu) [inline android/app/src/main/res/raw/app.json]: Github is warning me about there not being a newline at the end of this file, not sure if you intended to not leave a newline here.

File: src/screens/Resources/HomeScreen.js [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+import { collections } from '../../constants';
+import { HomeIcons } from './ResourcesComponents';
+import useLoadScreen from './LoadScreen';
+
+export default function HomeScreen(props) {
+  const { navigation } = props;
+  const [loadingNextScreen, err, loadScreen] = useLoadScreen(navigation);
+  const [errorMessage, setErrorMessage] = useState(err);
+  const [snapshot, setSnapshot] = useState(null);
+  const [initialLoad, setInitialLoad] = useState(true);
```
Comment (enmingliu) [inline src/screens/Resources/HomeScreen.js]: Maybe its better to set this state to false at the start and set it to true when we actually query from firestore and have it represent just `loading` in case we have something in the future that reloads the props apart from just opening the app.

File: src/screens/Resources/LoadScreen.js [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+  const [errorMessage, setErrorMessage] = useState(null);
+
+  async function firestoreQuery(collection, id) {
+    return firestore()
+      .collection(collection)
+      .where('parent', '==', id)
+      .get();
+  }
+
+  async function loadScreen(id, title) {
```
Comment (enmingliu) [inline src/screens/Resources/LoadScreen.js:18]: I really like the idea of having a function to handle querying & loading into arbitrary subcategories/subcategory lists, but is it possible for us to be in a loading state in the newly pushed screen (ie., showing nothing but an activity indicator) instead of waiting for the query to complete in the old screen? I ask this because having the activity indicator shifting ui elements around in the old screen while locking any ui input isn't very natural, especially if the query fails for some reason. (vs being in a new screen and showing an error message there)

File: src/screens/Resources/LoadScreen.js [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+    setLoading(true);
+
+    try {
```
Comment (enmingliu) [inline src/screens/Resources/LoadScreen.js:21]: Is there a specific reason we're using a try/catch block here instead of a promise off of firestore().collection().get()?

File: src/screens/Resources/ResourcesComponents/SingleItemList.js [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+            More&nbsp;
+            <Icon name="chevron-right" />
+          </Text>
+        </View>
+      </Collapsible>
+    </View>
+  );
+}
+
+const SingleItemStyles = StyleSheet.create({
```
Comment (enmingliu) [inline src/screens/Resources/ResourcesComponents/SingleItemList.js]: I'm not sure exactly what causes this problem, but the left/right margins for each list item are different when the accordion is extended vs not. (at least on iOS simulator, transportation -> big red bus) Is this intended?

File: src/screens/Resources/ResourcesComponents/MultipleItemList.js [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+      sections={snapshot.docs.map((doc) => ({ data: doc.data(), id: doc.id }))}
+      activeSections={activeSections}
+      renderHeader={renderHeader}
+      renderContent={renderContent}
+      onChange={updateSections}
+    />
+  );
+}
+
+const MultipleItemStyles = StyleSheet.create({
```
Comment (enmingliu) [inline src/screens/Resources/ResourcesComponents/MultipleItemList.js]: Also, it feels like a lot of styles are shared between MultipleItemStyles and SingleItemStyles, it would be great if we could reduce code reuse.

**Review by enmingliu — CHANGES_REQUESTED** [review]

**Comment by enmingliu** [comment]
This is a huge change, but it looks really good! Great work!

File: src/screens/Resources/LoadScreen.js (hunk 15 above)
Comment (Derek-Chu) [reply src/screens/Resources/LoadScreen.js:21]: we do that because otherwise it gets kinda messy and unclear bc i have to keep changing .then() statements. the async await generally makes it a bit clearer.

File: src/screens/Resources/ResourcesComponents/SingleItemList.js (hunk 16 above)
Comment (Derek-Chu) [reply src/screens/Resources/ResourcesComponents/SingleItemList.js]: interesting but i cant test since i dont have an ios environment :(

File: src/screens/Resources/LoadScreen.js (hunk 14 above)
Comment (Derek-Chu) [reply src/screens/Resources/LoadScreen.js:18]: to my knowledge not really. the reason why we handled it here is because we don't know whether we should be routing to a subcategories screen or a resources item list screen. because of that we need to query firebase to see which one we need to go to. and since we query firebase, i took the liberty of just extracting the snapshot and passing it to the next screen anyways. i'm open to any suggestions for how to change this tho!

Comment (enmingliu) [reply src/screens/Resources/LoadScreen.js:18]: I think we could have a staging screen whose only purpose is to indicate that its in the process of loading the subcategories/resource item list, like how the facebook app shows grey boxes before all the content loads, since I think the issue is mainly with having the activity indicator & error message on the previous screen.

File: src/screens/Resources/LoadScreen.js (hunk 15 above)
Comment (enmingliu) [reply src/screens/Resources/LoadScreen.js:21]: Sounds good!

File: src/screens/Resources/ResourcesComponents/SingleItemList.js (hunk 16 above)
Comment (enmingliu) [reply src/screens/Resources/ResourcesComponents/SingleItemList.js]: Ok, I don't think this is too big a deal, I can probably take another look at it after we merge this pr.

File: src/screens/Resources/ResourcesComponents/MultipleItemList.js (hunk 17 above)
Comment (Derek-Chu) [reply src/screens/Resources/ResourcesComponents/MultipleItemList.js]: I'll do this for now but I think the styles can be looked at more thoroughly in a future pr as well since i think this one is mostly a functionality-based one

File: src/screens/Resources/LoadScreen.js (hunk 14 above)
Comment (Derek-Chu) [reply src/screens/Resources/LoadScreen.js:18]: hmm i see and then display the error msg and the activity indicator in this staging screen?

Comment (enmingliu) [reply src/screens/Resources/LoadScreen.js:18]: Yup, exactly

**Review by enmingliu — APPROVED** [review]

**Review by vfcheung — APPROVED** [review]
